File: CPP/ByRound/1913/1913_F_Palindromic_Problem.cpp
// Problem: CF 1913 F - Palindromic Problem
// https://codeforces.com/contest/1913/problem/F

/*
 * Problem: F. Palindromic Problem
 * 
 * Algorithm: Manacher's Algorithm for finding palindromes, followed by a dynamic programming approach
 *            to determine the best character replacement to maximize palindromic substrings.
 * 
 * Time Complexity: O(n) for Manacher's algorithm + O(n^2) for DP = O(n^2) in worst case
 * Space Complexity: O(n) for storing hash values, palindrome lengths, and DP table
 * 
 * Approach:
 * 1. Use Manacher's algorithm to compute the longest palindromic substring centered at each point.
 * 2. For each position, calculate how changing that character affects palindromic counts.
 * 3. Use a DP table to track the benefit of changing a character to each possible letter.
 * 4. Select the lexicographically smallest string with maximum palindromes.
 */

#include <stdio.h>
#include <algorithm>
#include <iostream>
#include <string>

using namespace std;
#define MAXN 300000
long long dp[MAXN + 5][30];
unsigned long long h1[MAXN + 5], h2[MAXN + 5], b[MAXN + 5];
int base = 131;
string s;
int n;
unsigned long long solve1(int x, int y) {
  return h1[y] - h1[x - 1] * b[y - x + 1];
}
unsigned long long solve2(int x, int y) {
  x = n - x + 1, y = n - y + 1;
  return h2[y] - h2[x - 1] * b[y - x + 1];
}
int d1[MAXN + 5], d2[MAXN + 5];
long long sum[MAXN + 5], sum1[MAXN + 5];
void manacher() {
  for (int i = 0, l = 0, r = -1; i < n; i++) {
    int k = (i > r) ? 1 : min(d1[l + r - i], r - i + 1);
    while (0 <= i - k && i + k < n && s[i - k] == s[i + k])
      k++;
    d1[i] = k--;
    if (i + k > r) {
      l = i - k;
      r = i + k;
    }
  }
  for (int i = 0, l = 0, r = -1; i < n; i++) {
    int k = (i > r) ? 0 : min(d2[l + r - i + 1], r - i + 1);
    while (0 <= i - k - 1 && i + k < n && s[i - k - 1] == s[i + k])
      k++;
    d2[i] = k--;
    if (i + k > r) {
      l = i - k - 1;
      r = i + k;
    }
  }
  for (int i = n; i >= 1; i--) {
    d1[i] = d1[i - 1];
    d2[i] = d2[i - 1];
  }
}
int check(int x, int y) {
  if (x < 1 || y > n)
    return 0;
  int l = 0, r = min(x, n - y + 1);
  while (l <= r) {
    int mid = (l + r) >> 1;
    if (solve1(x - mid + 1, x) == solve2(y + mid - 1, y))
      l = mid + 1;
    else
      r = mid - 1;
  }
  return r;
}
int main() {
  cin >> n >> s;
  b[0] = 1;
  for (int i = 1; i <= n; i++) {
    h1[i] = h1[i - 1] * base + s[i - 1] - 'a' + 1;
    h2[i] = h2[i - 1] * base + s[n - i] - 'a' + 1;
    b[i] = b[i - 1] * base;
  }
  manacher();
  long long ans1 = 0, ans = 0;
  for (int i = 1; i <= n; i++)
    ans1 += d1[i] + d2[i];
  for (int i = 1; i <= n; i++) {
    if (d1[i] > 1) {
      int l = i - d1[i] + 1, r = i + d1[i] - 1;
      sum[l]++, sum[i]--, sum1[i] -= i - l;
      sum[i + 1]--, sum[r + 1]++, sum1[r + 1] += r - i;
      sum1[i + 1] += r - i + 1, sum1[r + 1] += -(r - i + 1);
    }
    if (d2[i] == 0)
      continue;
    int l = i - d2[i], r = i + d2[i] - 1;
    sum[l]++, sum[i]--, sum1[i] -= i - l;
    sum[i]--, sum[r + 1]++, sum1[r + 1] += r - i + 1;
    sum1[i] += r - i + 2, sum1[r + 1] += -(r - i + 2);
  }
  for (int i = 1; i <= n; i++)
    sum[i] = sum[i] + sum[i - 1];
  for (int i = 1; i <= n; i++)
    sum[i] = sum[i] + sum1[i];
  for (int i = 1; i <= n; i++)
    sum[i] = sum[i] + sum[i - 1];
  for (int i = 1; i <= n; i++)
    for (int j = 0; j < 26; j++)
      dp[i][j] = -sum[i];
  for (int i = 1; i <= n; i++) {
    int l = i - d1[i], r = i + d1[i];
    if (l >= 1 && r <= n) {
      int len = check(l - 1, r + 1);
      dp[l][s[r - 1] - 'a'] += len + 1;
      dp[r][s[l - 1] - 'a'] += len + 1;
    }
    l = i - d2[i] - 1, r = i + d2[i];
    if (l >= 1 && r <= n) {
      int len = check(l - 1, r + 1);
      dp[l][s[r - 1] - 'a'] += len + 1;
      dp[r][s[l - 1] - 'a'] += len + 1;
    }
  }
  int x = 0, y = 0;
  for (int i = 1; i <= n; i++) {
    for (int j = 25; j >= 0; j--) {
      if (dp[i][j] >= ans)
        ans = dp[i][j];
    }
  }
  int f = 1;
  for (int i = 1; i <= n; i++) {
    for (int j = 25; j >= 0; j--) {
      if (dp[i][j] == ans && j < s[i - 1] - 'a') {
        x = i, y = j;
        f = 0;
      }
      if (dp[i][j] == ans)
        x = i, y = j;
    }
    if (f == 0)
      break;
  }
  printf("%lld\n", ans1 + ans);
  if (x)
    s[x - 1] = char(y + 'a');
  cout << s;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/F/1913_F_Palindromic_Problem.py
# Problem: CF 1913 F - Palindromic Problem
# https://codeforces.com/contest/1913/problem/F

"""
Algorithm: Palindromic Problem
Approach:
- Use Manacher's algorithm to find all palindromes in the string.
- Calculate the number of palindromic substrings for each position.
- For each position, determine the optimal character to replace it with to maximize the number of palindromes.
- If there is a tie in the number of palindromes, choose the lexicographically smallest string.

Time Complexity: O(n log n)
Space Complexity: O(n)

Manacher's algorithm implementation is used to efficiently find all palindromic substrings in O(n) time.
The rest of the code manipulates these palindromes to find the best replacement strategy.
"""

import io
import os
import random
import sys

input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline


def manacher(s):
    i, j = 0, 0
    n = len(s)
    ans = [0] * n
    while i < n:
        while i - j >= 0 and i + j < n and s[i - j] == s[i + j]:
            j += 1
        ans[i] = j
        k = 1
        while i - k >= 0 and k + ans[i - k] < j:
            ans[i + k] = ans[i - k]
            k += 1
        i += k
        j -= k
    return ans


def f(u, v, w):
    return u << 38 ^ v << 19 ^ w


def random_mod():
    mod, p = 2 * random.randint(3 * pow(10, 8), 5 * pow(10, 8)) + 1, 1
    while not mod % p:
        mod, p = mod + 2, 3
        while p * p < mod and mod % p:
            p += 2
    return mod


n = int(input())
s = list(input().rstrip())
s = [i - 96 for i in s]
u = [0] * (2 * n + 1)
for i in range(n):
    u[2 * i + 1] = s[i]
p = manacher(u)
p = [i >> 1 for i in p[1:-1]]
c = [0] * (n + 2)
for i in range(n):
    l, m, r = i - p[2 * i] + 1, i + 1, i + p[2 * i] + 1
    c[l], c[m], c[r] = c[l] + 1, c[m] - 2, c[r] + 1
for i in range(n - 1):
    if not p[2 * i + 1]:
        continue
    l, m1, m2, r = i - p[2 * i + 1] + 1, i + 1, i + 2, i + p[2 * i + 1] + 2
    c[l], c[m1], c[m2], c[r] = c[l] + 1, c[m1] - 1, c[m2] - 1, c[r] + 1
for _ in range(2):
    for i in range(1, n + 2):
        c[i] += c[i - 1]
for i in range(n):
    c[i] -= p[2 * i]
x = []
for i in range(2 * n - 1):
    l, r = (i >> 1) - p[i], (i >> 1) + p[i] + (i & 1)
    if 0 <= l and r < n:
        x.append(f(l << 5 ^ s[r], l, r))
        x.append(f(r << 5 ^ s[l], l, r))
m1, m2 = random_mod(), random_mod()
p1, p2 = [1] * (2 * n + 1), [1] * (2 * n + 1)
h1, h2 = [0] * (2 * n + 1), [0] * (2 * n + 1)
for i in range(1, 2 * n + 1):
    p1[i] = 29 * p1[i - 1] % m1
    p2[i] = 29 * p2[i - 1] % m2
ss = s + s[::-1]
for i in range(2 * n):
    h1[i + 1] = h1[i] + p1[i] * ss[i]
    h2[i + 1] = h2[i] + p2[i] * ss[i]
pow2 = [1]
for _ in range(n.bit_length()):
    pow2.append(2 * pow2[-1])
s.append(0)
la, ma, k, cnt = n << 5, 0, n << 5, 0
for i in sorted(x):
    u, l, r = i >> 38, (i >> 19) & 0x7FFFF, i & 0x7FFFF
    v = 0
    if l and (r + 1) ^ n and s[l - 1] == s[r + 1]:
        d = 2 * n - (l + r + 1)
        for j in pow2[::-1]:
            l0, l1, r0, r1 = l - v - j, l - 1, r + 1, r + v + j
            if not 0 <= l0 or not r1 < n:
                continue
            hl = (h1[l1 + 1] - h1[l0]) % m1 * p1[d] % m1
            hr = (h1[2 * n - r0] - h1[2 * n - r1 - 1]) % m1
            if hl ^ hr:
                continue
            hl = (h2[l1 + 1] - h2[l0]) % m2 * p2[d] % m2
            hr = (h2[2 * n - r0] - h2[2 * n - r1 - 1]) % m2
            if hl ^ hr:
                continue
            v += j
    v += 1
    if la ^ u:
        y = cnt - c[la >> 5]
        if ma < y or (ma == y and s[k >> 5] <= k & 31 and (la >> 5) ^ (k >> 5)):
            ma, k = y, la
        cnt = 0
    la, cnt = u, cnt + v
y = cnt - c[la >> 5]
if ma < y or (ma == y and s[k >> 5] <= k & 31 and (la >> 5) ^ (k >> 5)):
    ma, k = y, la
ans = [chr(i + 96) for i in s[:-1]]
if (k >> 5) ^ n:
    ans[k >> 5] = chr((k & 31) + 96)
if n == 1:
    ans = ["a"]
m = sum(p) + ma
print(m)
sys.stdout.write("".join(ans))


# https://github.com/VaHiX/CodeForces/